warp10/src/main/java/io/warp10/script/formatted/FormattedWarpScriptFunction.java

```diff
@@ -0,0 +1,305 @@
+//
+//   Copyright 2019  SenX S.A.S.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+
+package io.warp10.script.formatted;
+
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Do the extraction of arguments from the stack in a formatted manner.
+ * Arguments inside collections still have to be handled manually.
+ *
+ * Alternatively, a Map that contains all arguments and optional arguments can be provided on top of the stack.
+ *
+ * See FormattedWarpScriptFunctionTest for an example of child class.
+ */
+public abstract class FormattedWarpScriptFunction extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  private final StringBuilder docstring;
+  private final List<String> unitTests;
+
+  public static class Arguments {
+    private final List<ArgumentSpecification> args;
+    private final List<ArgumentSpecification> optArgs;
+
+    private Arguments(List<ArgumentSpecification> args, List<ArgumentSpecification> optArgs) {
+      this.args = args;
+      this.optArgs = optArgs;
+    }
+
+    public List<ArgumentSpecification> getArgsCopy() {
+      return new ArrayList<>(args);
+    }
+
+    public List<ArgumentSpecification> getOptArgsCopy() {
+      return new ArrayList<>(optArgs);
+    }
+  }
+
+  public static class ArgumentsBuilder {
+    private final List<ArgumentSpecification> args;
+    private final List<ArgumentSpecification> optArgs;
+
+    public ArgumentsBuilder() {
+      args = new ArrayList<>();
+      optArgs = new ArrayList<>();
+    }
+
+    public ArgumentsBuilder addArgument(Class<?> clazz, String name, String doc) {
+      args.add(new ArgumentSpecification(clazz, name, doc));
+      return this;
+    }
+
+    public ArgumentsBuilder addOptionalArgument(Class<?> clazz, String name, Object defaultValue, String doc) {
```

Comment:
Can you refactor this so the first 3 parameters are identical to the other signature, this will make it easier to add a default value afterwards.